src/codecatalyst/commands.ts

```diff
@@ -135,13 +136,43 @@ function createClientInjector(authProvider: CodeCatalystAuthenticationProvider):
 
         await authProvider.restore()
         const conn = authProvider.activeConnection ?? (await authProvider.promptNotConnected())
-        const client = await createClient(conn)
+        const authenticatedConn = await authenticateConnection(conn, authProvider.auth)
+        const client = await createClient(authenticatedConn)
         telemetry.record({ userId: client.identity.id })
 
         return command(client, ...args)
     }
 }
 
+/**
+ * Returns a connection that is ensured to be authenticated.
+ *
+ * Provides the user the ability to re-authenticate if needed,
+ * otherwise throwing an error.
+ */
+async function authenticateConnection(conn: SsoConnection, auth: Auth): Promise<SsoConnection> {
+    if (auth.getConnectionState(conn) === 'valid') {
+        return conn
+    }
+
+    // Have user try to log in
+    const loginMessage = localize('aws.auth.invalidConnection', 'Connection is invalid or expired, login again?')
+    const result = await vscode.window.showErrorMessage(loginMessage, 'Login')
+
+    if (result !== 'Login') {
+        throw new ToolkitError('User cancelled login.', { cancelled: true, code: invalidConnection })
+    }
+
+    conn = await auth.reauthenticate(conn)
+
+    // Log in attempt failed
+    if (auth.getConnectionState(conn) !== 'valid') {
+        throw new ToolkitError('Connection could not be authenticated.', { code: invalidConnection })
```

Comment:
this is user-facing, let's use more intuitive wording. "Failed to connect", or "Connection failed", or...

```suggestion
        throw new ToolkitError('Could not connect.', { code: invalidConnection })
```